[ASSB-1334] Detect Actual Changes and Display Changed Flag at field, sub field, protocol , sections

mohomeoffice opened the pull request:
This PR implements the following changes:
- Detect actual changes at both field and sub-field levels.
- Display the changed flag only when genuine modifications occur.
- Updated the following components: application-summary, checkbox, editor, review-field, review, static-section, protocol-sections, steps, and reducers.

jeff-horton-ho-sas commented on client/components/application-summary.js:
```diff
@@ -20,6 +20,42 @@ import Submit from './submit';
 import { selector } from './sync-handler';
 import HoldingPage from './holding-page';
 
+
+/**
+ * Normalise a given value into a consistent string format.
+ * 
+ * - If the value is `null` or `undefined`, it returns an empty string.
+ * - If the value is an object, it converts it into a JSON string.
+ * - For all other types, it converts the value into a string and trims any extra whitespace.
+ *
+ * @param {any} value - The input value to normalise.
+ * @returns {string} - A normalised string representation of the input value.
+ */
+function normalizeValue(value) {
+  if (value === null || value === undefined) {
+    return ""; // Return an empty string for null or undefined values.
+  }
+  if (typeof value === "object") {
+    return JSON.stringify(value); // Convert objects to their JSON string representation.
+  }
+  return String(value).trim(); // Convert other types to string and trim whitespace.
+}
+
+/**
+ * Sanitise a given value by normalising it and removing double quotes.
+ *
+ * - First, the value is normalised using `normalizeValue`, ensuring it is a consistent string.
+ * - Then, any double quotes (`"`) in the string are removed using a regular expression.
+ *
+ * @param {any} value - The input value to sanitise.
+ * @returns {string} - A sanitised string with double quotes removed.
+ */
+function sanitizeValue(value) {
+  return normalizeValue(value).replace(/["]+/g, ""); // Remove all double quotes from the normalised value.
+}
```
See comment on other PR https://github.com/UKHomeOffice/asl-projects/pull/974/files#r1931976389

jeff-horton-ho-sas commented on client/components/review.js:
```diff
@@ -11,9 +11,99 @@ import { Markdown } from '@ukhomeoffice/asl-components';
+    // console.log(`Normalized Current Value:`, normalizedCurrentValue);
+    // console.log(`Value for this field ${name}:`, this.props.values[name]);
+
+    // Detect changes by comparing normalized values
+    let hasChange = !isEqual(normalizedStoredValue, normalizedCurrentValue);
+    //console.log(`Change Detected for ${name}:`, hasChange);
+
+    // Add logic for species-based dynamic fields
+    const species = values?.species || [];
+    if (species.length === 0) {
+       // console.error('Species array is empty or undefined.');
+    } else {
+        //console.log('Checking species-based dynamic fields...');
+        species.forEach(speciesName => {
+            const fieldName = `reduction-quantities-${speciesName}`;
+            const speciesStoredValue = values.storedValue?.[fieldName] || null;
+            const speciesCurrentValue = values[fieldName] || null;
+
+            // console.log(`Field Name: ${fieldName}`);
+            // console.log(`Stored Value:`, speciesStoredValue);
+            // console.log(`Current Value:`, speciesCurrentValue);
+
+            if (!isEqual(normalizeValue(speciesStoredValue), normalizeValue(speciesCurrentValue))) {
+                //console.log(`Change Detected for ${fieldName}: true`);
+                hasChange = true; // Any change in dynamic fields sets hasChange to true
+            } else {
+                //console.log(`Change Detected for ${fieldName}: false`);
+            }
+        });
```
See comment on other PR https://github.com/UKHomeOffice/asl-projects/pull/974/files#r1932096882

jeff-horton-ho-sas commented on client/components/review.js:
```diff
@@ -35,24 +125,24 @@ class Review extends React.Component {
       return <Comments field={`${this.props.prefix || ''}${this.props.name}`} collapsed={!this.props.readonly} />;
     }
 
+   // console.log('FIELD CALLED HERE');
```
```suggestion
```

jeff-horton-ho-sas commented on client/pages/sections/protocols/sections.js:
```diff
@@ -106,13 +110,68 @@ const getBadges = (section, newComments, values) => {
+      fieldValue = rawValue || null;
+    }
+
+    console.log(`Field ${index + 1}: ${field.name} = ${fieldValue}`);
```
```suggestion
```

jeff-horton-ho-sas replied:
These logs are going to be very noisy in opensearch when doing server-side rendering of the page, and could potentially be logging PII which would be a security risk

jeff-horton-ho-sas commented on client/pages/sections/protocols/sections.js:
```diff
@@ -106,13 +110,68 @@ const getBadges = (section, newComments, values) => {
+      });
+    }
+  });
+
+  console.log("Fields with values:", fieldsWithValues);
+  console.log("Fields without values:", fieldsWithoutValues);
```
```suggestion
```

jeff-horton-ho-sas replied:
These logs are going to be very noisy in opensearch when doing server-side rendering of the page, and could potentially be logging PII which would be a security risk

jeff-horton-ho-sas commented on client/components/review.js:
```diff
@@ -11,9 +11,99 @@ import { Markdown } from '@ukhomeoffice/asl-components';
+        name === 'duration' ? normalizeDuration(storedValue) : storedValue;
+
+    // Determine the actual current value for specific fields
+    const actualCurrentValue = currentValue !== undefined && currentValue !== null
+        ? currentValue
+        : values[name] !== undefined
+        ? values[name]
+        : null; // Fallback to `null` if no value is provided
```
This is hard to read and can use the [nullish coalescing operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing) to make it clearer
```suggestion
    const actualCurrentValue = currentValue ?? values[name] ?? null
```

mohomeoffice replied:
Jeff, I initially used the nullish operator for readability, but the explicit ternary check ensures falsy values like 0, "", and false are preserved, preventing unintended fallbacks, so I had to keep it this way.

jeff-horton-ho-sas replied:
So does `??` - it only uses the RHS if the LHS is `null` or `undefined`. `||` has the behaviour you're describing, but that isn't what i'm suggesting

jeff-horton-ho-sas commented on client/components/review.js:
```diff
@@ -11,9 +11,99 @@ import { Markdown } from '@ukhomeoffice/asl-components';
+        if (value && typeof value === 'object' && ('years' in value || 'months' in value)) {
+            // Normalize duration objects specifically
+            return {
+                years: value.years !== undefined && value.years !== null ? value.years : 5, // Default to 5 if undefined or null
+                months: value.months !== undefined && value.months !== null ? value.months : 0 // Default to 0 if undefined or null
```
```suggestion
                years: value.years ?? 5,
                months: value.months ?? 0
```

jeff-horton-ho-sas commented on client/components/application-summary.js:
```diff
@@ -205,6 +245,117 @@ const ApplicationSummary = () => {
+const hasChangedFields = (fields, currentValues, initialValues) => {
+  return fields.some(field => {
+    // Normalize and sanitize current and initial values
+    const currentValue = sanitizeValue(currentValues[field]);
+    const initialValue = sanitizeValue(initialValues[field]);
```
This will ensure that the currentValue and initialValue are strings, so the special handling for arrays and booleans will never be used and can be removed.

jeff-horton-ho-sas commented on client/components/review.js:
```diff
@@ -11,9 +11,99 @@ import { Markdown } from '@ukhomeoffice/asl-components';
+        return { years: 5, months: 0 }; // Default for completely missing duration
+    };
+
+    const { name, storedValue, currentValue, values } = this.props;
```
Can these be passed into `hasDatabaseChanges` as parameters?

Reasoning:
* This would allow hasDatabaseChanges to have unit tests independent of the component
* We should be making it as simple as possible to switch from class to function components, where the props are passed in as arguments and are not accessible in this way.

mohomeoffice replied:
Passing a parameter would improve the code design. I appreciate the feedback and will incorporate it in the second part of this task.

jeff-horton-ho-sas commented on client/components/application-summary.js:
```diff
@@ -205,6 +245,117 @@ const ApplicationSummary = () => {
+          sanitizedSaved !== sanitizedCurrent ||
+          sanitizedSaved !== sanitizedInitial
+        ) {
+          console.log(`  Field "${key}" has changed.`);
```
```suggestion
```

jeff-horton-ho-sas reviewed (changes requested):
Mostly looks good, thanks.

* Remove debug / commented out console logs
* Some refactorings for readability
* It would be good to see some test cases added for hasDatabaseChanges, hasSectionChangedDeep, and hasChangedFields. There are some existing tests in https://github.com/UKHomeOffice/asl-projects/tree/master/test/specs as an example.

mohomeoffice commented:
[like]  Mokter Hossain reacted to your message:
________________________________
From: Jeff Horton ***@***.***>
Sent: Wednesday, January 29, 2025 9:40:17 AM
To: UKHomeOffice/asl-projects ***@***.***>
Cc: Mokter Hossain ***@***.***>; Author ***@***.***>
Subject: Re: [UKHomeOffice/asl-projects] [ASSB-1334] Detect Actual Changes and Display Changed Flag at field, sub field, protocol , sections (PR #973)


@jeff-horton-ho-sas requested changes on this pull request.

Mostly looks good, thanks.

  *   Remove debug / commented out console logs
  *   Some refactorings for readability
  *   It would be good to see some test cases added for hasDatabaseChanges, hasSectionChangedDeep, and hasChangedFields. There are some existing tests in https://github.com/UKHomeOffice/asl-projects/tree/master/test/specs as an example.

________________________________

In client/components/application-summary.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933486574>:

> +/**
+ * Normalise a given value into a consistent string format.
+ *
+ * - If the value is `null` or `undefined`, it returns an empty string.
+ * - If the value is an object, it converts it into a JSON string.
+ * - For all other types, it converts the value into a string and trims any extra whitespace.
+ *
+ * @param {any} value - The input value to normalise.
+ * @returns {string} - A normalised string representation of the input value.
+ */
+function normalizeValue(value) {
+  if (value === null || value === undefined) {
+    return ""; // Return an empty string for null or undefined values.
+  }
+  if (typeof value === "object") {
+    return JSON.stringify(value); // Convert objects to their JSON string representation.
+  }
+  return String(value).trim(); // Convert other types to string and trim whitespace.
+}
+
+/**
+ * Sanitise a given value by normalising it and removing double quotes.
+ *
+ * - First, the value is normalised using `normalizeValue`, ensuring it is a consistent string.
+ * - Then, any double quotes (`"`) in the string are removed using a regular expression.
+ *
+ * @param {any} value - The input value to sanitise.
+ * @returns {string} - A sanitised string with double quotes removed.
+ */
+function sanitizeValue(value) {
+  return normalizeValue(value).replace(/["]+/g, ""); // Remove all double quotes from the normalised value.
+}


See comment on other PR https://github.com/UKHomeOffice/asl-projects/pull/974/files#r1931976389

________________________________

In client/components/application-summary.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933486847>:

> +        // Compare values and log debugging information
+        // console.log(`Field: ${key}`);
+        // console.log(`  Saved: ${sanitizedSaved}`);
+        // console.log(`  Current: ${sanitizedCurrent}`);
+        // console.log(`  Initial: ${sanitizedInitial}`);


⬇️ Suggested change

-        // Compare values and log debugging information
-        // console.log(`Field: ${key}`);
-        // console.log(`  Saved: ${sanitizedSaved}`);
-        // console.log(`  Current: ${sanitizedCurrent}`);
-        // console.log(`  Initial: ${sanitizedInitial}`);


________________________________

In client/components/application-summary.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933487176>:

> +      // console.log(`Field: ${field}`);
+      // console.log(`Current (Array):`, currentArray);
+      // console.log(`Initial (Array):`, initialArray);
+      // console.log(`Changed (Array):`, changed);


⬇️ Suggested change

-      // console.log(`Field: ${field}`);
-      // console.log(`Current (Array):`, currentArray);
-      // console.log(`Initial (Array):`, initialArray);
-      // console.log(`Changed (Array):`, changed);


________________________________

In client/components/application-summary.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933487438>:

> +      // console.log(`Field: ${field}`);
+      // console.log(`Current (Boolean):`, currentValue);
+      // console.log(`Initial (Boolean):`, initialValue);
+      // console.log(`Changed (Boolean):`, changed);


⬇️ Suggested change

-      // console.log(`Field: ${field}`);
-      // console.log(`Current (Boolean):`, currentValue);
-      // console.log(`Initial (Boolean):`, initialValue);
-      // console.log(`Changed (Boolean):`, changed);


________________________________

In client/components/application-summary.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933487762>:

> +
+    // console.log(`Field: ${field}`);
+    // console.log(`Current (Sanitized): "${currentValue}"`);
+    // console.log(`Initial (Sanitized): "${initialValue}"`);
+    // console.log(`Changed:`, changed);


⬇️ Suggested change

-
-    // console.log(`Field: ${field}`);
-    // console.log(`Current (Sanitized): "${currentValue}"`);
-    // console.log(`Initial (Sanitized): "${initialValue}"`);
-    // console.log(`Changed:`, changed);


________________________________

In client/components/review.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933488306>:

> +    // console.log(`Field Name: ${name}`);
+    // console.log(`Stored Value:`, adjustedStoredValue);
+    // console.log(`Actual Current Value:`, actualCurrentValue);
+    // console.log(`Normalized Stored Value:`, normalizedStoredValue);
+    // console.log(`Normalized Current Value:`, normalizedCurrentValue);
+    // console.log(`Value for this field ${name}:`, this.props.values[name]);


⬇️ Suggested change

-    // console.log(`Field Name: ${name}`);
-    // console.log(`Stored Value:`, adjustedStoredValue);
-    // console.log(`Actual Current Value:`, actualCurrentValue);
-    // console.log(`Normalized Stored Value:`, normalizedStoredValue);
-    // console.log(`Normalized Current Value:`, normalizedCurrentValue);
-    // console.log(`Value for this field ${name}:`, this.props.values[name]);


________________________________

In client/components/review.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933492672>:

> +    // Adjust storedValue for duration fields
+    const adjustedStoredValue =
+        name === 'duration' ? normalizeDuration(storedValue) : storedValue;


Normalise is called for durations on lines 61/63 so this is redundant

⬇️ Suggested change

-    // Adjust storedValue for duration fields
-    const adjustedStoredValue =
-        name === 'duration' ? normalizeDuration(storedValue) : storedValue;


________________________________

In client/components/review.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933496959>:

> +    // Detect changes by comparing normalized values
+    let hasChange = !isEqual(normalizedStoredValue, normalizedCurrentValue);
+    //console.log(`Change Detected for ${name}:`, hasChange);
+
+    // Add logic for species-based dynamic fields
+    const species = values?.species || [];
+    if (species.length === 0) {
+       // console.error('Species array is empty or undefined.');
+    } else {
+        //console.log('Checking species-based dynamic fields...');
+        species.forEach(speciesName => {
+            const fieldName = `reduction-quantities-${speciesName}`;
+            const speciesStoredValue = values.storedValue?.[fieldName] || null;
+            const speciesCurrentValue = values[fieldName] || null;
+
+            // console.log(`Field Name: ${fieldName}`);
+            // console.log(`Stored Value:`, speciesStoredValue);
+            // console.log(`Current Value:`, speciesCurrentValue);
+
+            if (!isEqual(normalizeValue(speciesStoredValue), normalizeValue(speciesCurrentValue))) {
+                //console.log(`Change Detected for ${fieldName}: true`);
+                hasChange = true; // Any change in dynamic fields sets hasChange to true
+            } else {
+                //console.log(`Change Detected for ${fieldName}: false`);
+            }
+        });


See comment on other PR https://github.com/UKHomeOffice/asl-projects/pull/974/files#r1932096882

________________________________

In client/components/review.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933497441>:

>
     if (this.props.type === 'comments-only' && showComments) {
       return <Comments field={`${this.props.prefix || ''}${this.props.name}`} collapsed={!this.props.readonly} />;
     }

+   // console.log('FIELD CALLED HERE');


⬇️ Suggested change

-   // console.log('FIELD CALLED HERE');


________________________________

In client/pages/sections/protocols/sections.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933498007>:

> +      // Handle structured data
+      if (rawValue.object) {
+        fieldValue = rawValue.object; // Extract `object` key if available
+      } else if (Array.isArray(rawValue)) {
+        fieldValue = rawValue.join(', '); // Join array elements
+      } else if (rawValue.document) {
+        fieldValue = JSON.stringify(rawValue.document); // Handle `document` key
+      } else {
+        fieldValue = JSON.stringify(rawValue); // Convert object to string
+      }
+    } else {
+      // Handle simple values or fallback to default message
+      fieldValue = rawValue || null;
+    }
+
+    console.log(`Field ${index + 1}: ${field.name} = ${fieldValue}`);


⬇️ Suggested change

-    console.log(`Field ${index + 1}: ${field.name} = ${fieldValue}`);


________________________________

In client/pages/sections/protocols/sections.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933498456>:

> +
+  console.log("Fields with values:", fieldsWithValues);
+  console.log("Fields without values:", fieldsWithoutValues);


⬇️ Suggested change

-
-  console.log("Fields with values:", fieldsWithValues);
-  console.log("Fields without values:", fieldsWithoutValues);


________________________________

In client/components/review.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933503223>:

> +    const actualCurrentValue = currentValue !== undefined && currentValue !== null
+        ? currentValue
+        : values[name] !== undefined
+        ? values[name]
+        : null; // Fallback to `null` if no value is provided


This is hard to read and can use the nullish coalescing operator<https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing> to make it clearer

⬇️ Suggested change

-    const actualCurrentValue = currentValue !== undefined && currentValue !== null
-        ? currentValue
-        : values[name] !== undefined
-        ? values[name]
-        : null; // Fallback to `null` if no value is provided
+    const actualCurrentValue = currentValue ?? values[name] ?? null


________________________________

In client/components/review.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933509961>:

> +                years: value.years !== undefined && value.years !== null ? value.years : 5, // Default to 5 if undefined or null
+                months: value.months !== undefined && value.months !== null ? value.months : 0 // Default to 0 if undefined or null


⬇️ Suggested change

-                years: value.years !== undefined && value.years !== null ? value.years : 5, // Default to 5 if undefined or null
-                months: value.months !== undefined && value.months !== null ? value.months : 0 // Default to 0 if undefined or null
+                years: value.years ?? 5,
+                months: value.months ?? 0


________________________________

In client/components/application-summary.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933515158>:

> +    const currentValue = sanitizeValue(currentValues[field]);
+    const initialValue = sanitizeValue(initialValues[field]);


This will ensure that the currentValue and initialValue are strings, so the special handling for arrays and booleans will never be used and can be removed.

________________________________

In client/components/review.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933524542>:

> +        }
+        return value; // Return all other values as-is
+    };
+
+    const normalizeDuration = value => {
+        if (value && typeof value === 'object' && ('years' in value || 'months' in value)) {
+            // Normalize duration objects specifically
+            return {
+                years: value.years !== undefined && value.years !== null ? value.years : 5, // Default to 5 if undefined or null
+                months: value.months !== undefined && value.months !== null ? value.months : 0 // Default to 0 if undefined or null
+            };
+        }
+        return { years: 5, months: 0 }; // Default for completely missing duration
+    };
+
+    const { name, storedValue, currentValue, values } = this.props;


Can these be passed into hasDatabaseChanges as parameters?

Reasoning:

  *   This would allow hasDatabaseChanges to have unit tests independent of the component
  *   We should be making it as simple as possible to switch from class to function components, where the props are passed in as arguments and are not accessible in this way.

________________________________

In client/components/application-summary.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1933528475>:

> +
+        const sanitizedSaved = sanitizeValue(savedValue);
+        const sanitizedCurrent = sanitizeValue(currentValue);
+        const sanitizedInitial = sanitizeValue(initialValue);
+
+        // Compare values and log debugging information
+        // console.log(`Field: ${key}`);
+        // console.log(`  Saved: ${sanitizedSaved}`);
+        // console.log(`  Current: ${sanitizedCurrent}`);
+        // console.log(`  Initial: ${sanitizedInitial}`);
+
+        if (
+          sanitizedSaved !== sanitizedCurrent ||
+          sanitizedSaved !== sanitizedInitial
+        ) {
+          console.log(`  Field "${key}" has changed.`);


⬇️ Suggested change

-          console.log(`  Field "${key}" has changed.`);


—
Reply to this email directly, view it on GitHub<https://github.com/UKHomeOffice/asl-projects/pull/973#pullrequestreview-2580372260>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BLBLTTMWDPWYMPGSFMX6YKL2NCOYDAVCNFSM6AAAAABV75V2MSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKOBQGM3TEMRWGA>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

Please ensure that any communication with the Home Office is via an official account ending with digital.homeoffice.gov.uk, homeoffice.gov.uk or homeoffice.gsi.gov.uk. This email and any files transmitted with it are private and intended solely for the use of the individual or entity to whom they are addressed. If you have received this email in error please return it to the address it came from telling them it is not for you and then delete it from your system. Communications via the digital.homeoffice.gov.uk domain may be automatically logged, monitored and/or recorded for legal purposes. This email message has been swept for computer viruses.

jeff-horton-ho-sas commented:
It looks like there is a bug in eslint causing the linting to fail (https://drone-gh.acp.homeoffice.gov.uk/UKHomeOffice/asl-projects/4773/1/3). It looks like this might be fixed in more recent versions https://github.com/eslint/eslint/issues/14172

kiran-varma-home-office commented on client/components/application-summary.js:
```diff
@@ -20,6 +20,48 @@ import Submit from './submit';
+ * @param {any} value - The input value to normalise.
+ * @returns {string} - A normalised string representation of the input value.
+ */
+function normaliseValue(value) {
+
+if (value === null || value === undefined) {
+  return ""; // Return an empty string for null or undefined values.
+}
+
+if (typeof value === "object") {
+  return JSON.stringify(value); // Convert objects to their JSON string representation.
+}
+
+return String(value).trim(); // Convert other types to string and trim whitespace.
+
+}
```
@mohomeoffice - This function is exactly same as what you have in other PR. What is your plan?

mohomeoffice replied:
Hi Kiran, I have closed the other PR. The draft PR was meant to gather input from you and the team on test cases, not for a code review.

kiran-varma-home-office replied:
I can understand if you want to do what ever is comfortable for you to do your first PR but going forward its best to keep all work in 1 branch and open draft PR from the same branch. I have taken time to review and comment other PR. I hope its useful and not lost.

mohomeoffice commented:
[like]  Mokter Hossain reacted to your message:
________________________________
From: Kiran Varma ***@***.***>
Sent: Friday, January 31, 2025 10:01:15 AM
To: UKHomeOffice/asl-projects ***@***.***>
Cc: Mokter Hossain ***@***.***>; Mention ***@***.***>
Subject: Re: [UKHomeOffice/asl-projects] [ASSB-1334] Detect Actual Changes and Display Changed Flag at field, sub field, protocol , sections (PR #973)


@kiran-varma-home-office commented on this pull request.

________________________________

In client/components/application-summary.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1936979225>:

> +function normaliseValue(value) {
+
+if (value === null || value === undefined) {
+  return ""; // Return an empty string for null or undefined values.
+}
+
+if (typeof value === "object") {
+  return JSON.stringify(value); // Convert objects to their JSON string representation.
+}
+
+return String(value).trim(); // Convert other types to string and trim whitespace.
+
+}


I can understand if you want to do what ever is comfortable for you to do your first PR but going forward its best to keep all work in 1 branch and open draft PR from the same branch. I have taken time to review and comment other PR. I hope its useful and not lost.

—
Reply to this email directly, view it on GitHub<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1936979225>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BLBLTTJIHSJ2VRFCUED7ZC32NNCWXAVCNFSM6AAAAABV75V2MSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKOBWGE3TAOBUGA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Please ensure that any communication with the Home Office is via an official account ending with digital.homeoffice.gov.uk, homeoffice.gov.uk or homeoffice.gsi.gov.uk. This email and any files transmitted with it are private and intended solely for the use of the individual or entity to whom they are addressed. If you have received this email in error please return it to the address it came from telling them it is not for you and then delete it from your system. Communications via the digital.homeoffice.gov.uk domain may be automatically logged, monitored and/or recorded for legal purposes. This email message has been swept for computer viruses.

mohomeoffice commented:
[0]     Mokter Hossain reacted to your message:
________________________________
From: Kiran Varma ***@***.***>
Sent: Friday, January 31, 2025 10:01:15 AM
To: UKHomeOffice/asl-projects ***@***.***>
Cc: Mokter Hossain ***@***.***>; Mention ***@***.***>
Subject: Re: [UKHomeOffice/asl-projects] [ASSB-1334] Detect Actual Changes and Display Changed Flag at field, sub field, protocol , sections (PR #973)


@kiran-varma-home-office commented on this pull request.

________________________________

In client/components/application-summary.js<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1936979225>:

> +function normaliseValue(value) {
+
+if (value === null || value === undefined) {
+  return ""; // Return an empty string for null or undefined values.
+}
+
+if (typeof value === "object") {
+  return JSON.stringify(value); // Convert objects to their JSON string representation.
+}
+
+return String(value).trim(); // Convert other types to string and trim whitespace.
+
+}


I can understand if you want to do what ever is comfortable for you to do your first PR but going forward its best to keep all work in 1 branch and open draft PR from the same branch. I have taken time to review and comment other PR. I hope its useful and not lost.

—
Reply to this email directly, view it on GitHub<https://github.com/UKHomeOffice/asl-projects/pull/973#discussion_r1936979225>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BLBLTTJIHSJ2VRFCUED7ZC32NNCWXAVCNFSM6AAAAABV75V2MSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKOBWGE3TAOBUGA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Please ensure that any communication with the Home Office is via an official account ending with digital.homeoffice.gov.uk, homeoffice.gov.uk or homeoffice.gsi.gov.uk. This email and any files transmitted with it are private and intended solely for the use of the individual or entity to whom they are addressed. If you have received this email in error please return it to the address it came from telling them it is not for you and then delete it from your system. Communications via the digital.homeoffice.gov.uk domain may be automatically logged, monitored and/or recorded for legal purposes. This email message has been swept for computer viruses.

jeff-horton-ho-sas commented on client/components/application-summary.js:
```diff
@@ -205,6 +251,109 @@ const ApplicationSummary = () => {
+    const currentValue = cleanValue(currentValues[field]);
+    const initialValue = cleanValue(initialValues[field]);
+
+    // Special handling for array-based fields (e.g., checkboxes, multi-selects)
+    if (Array.isArray(currentValue) || Array.isArray(initialValue)) {
+      const currentArray = Array.isArray(currentValue) ? currentValue : [];
+      const initialArray = Array.isArray(initialValue) ? initialValue : [];
+      const changed = JSON.stringify(currentArray.sort()) !== JSON.stringify(initialArray.sort());
+
+      return changed;
+    }
+
+    // Special handling for boolean fields (e.g., checkboxes, radios)
+    if (typeof currentValue === "boolean" || typeof initialValue === "boolean") {
+      const changed = currentValue !== initialValue;
+
+      return changed;
+    }
```
After normalisation current value and initial value are always strings so this is redundant and misleading.
```suggestion
```

jeff-horton-ho-sas commented on client/components/review.js:
```diff
@@ -11,9 +11,106 @@ import { Markdown } from '@ukhomeoffice/asl-components';
+ * @returns {boolean} - `true` if there is a change in the data, otherwise `false`.
+ */
+hasDatabaseChange() {
+    const normaliseValue = value => {
+        if (value === null || value === undefined || (Array.isArray(value) && value.length === 0)) {
+            // Normalise `null`, `undefined`, and empty arrays to a consistent empty state
+            return null;
+        }
+        if (Array.isArray(value)) {
+            // Sort arrays to ensure order doesn't affect comparison
+            return value.sort();
+        }
+        if (typeof value === 'object' && value !== null) {
+            // Ensure consistent ordering of object keys for generic objects
+            return JSON.stringify(value, Object.keys(value).sort());
+        }
+        return value; // Return all other values as-is
+    };
```
This is different to `normaliseValue` in application-summary.js. We should probably extract both into a library method and have a consistent normalised form.

```js
        if (typeof value === 'object' && value !== null) {
            // Ensure consistent ordering of object keys for generic objects
            return JSON.stringify(value, Object.keys(value).sort());
        }
```
This is a shallow normalisation, any nested objects could still have inconsistent ordering of keys.

We already depend on `json-stable-stringify` via `shasum` to calculate if there's a mismatch between server and client versions when applying a patch. (https://github.com/UKHomeOffice/asl-projects/blob/master/client/actions/projects.js#L325-L331) we could add this as a direct dependency and use that rather than writing our own.

mohomeoffice replied:
I have used json-stable-stringify

jeff-horton-ho-sas commented on client/components/review.js:
```diff
@@ -11,9 +11,106 @@ import { Markdown } from '@ukhomeoffice/asl-components';
+        return value; // Return all other values as-is
+    };
+
+    const normaliseDuration = value => {
+    if (value && typeof value === 'object' && ('years' in value || 'months' in value)) {
+        // Normalise duration objects specifically
+        return {
+            years: value.years ?? 5, // Default to 5 if undefined or null
+            months: value.months ?? 0 // Default to 0 if undefined or null
+        };
+    }
+
+      // Default for completely missing duration
+      return { years: 5, months: 0 };
+  };
```
```suggestion
    const normaliseDuration = value => ({
        years: value?.years ?? 5, 
        months: value?.months ?? 0
    });
```

mohomeoffice replied:
code refactored

jeff-horton-ho-sas commented on client/components/review.js:
```diff
@@ -11,9 +11,106 @@ import { Markdown } from '@ukhomeoffice/asl-components';
+
+    const { name, storedValue, currentValue, values } = this.props;
+
+    // Adjust storedValue for duration fields
+    const adjustedStoredValue =
+        name === 'duration' ? normaliseDuration(storedValue) : storedValue;
```
Redundant because of lines 79 and 81.

```suggestion
```

mohomeoffice replied:
code refactored

jeff-horton-ho-sas commented on client/components/review.js:
```diff
@@ -11,9 +11,106 @@ import { Markdown } from '@ukhomeoffice/asl-components';
+    const normalisedCurrentValue =
+        name === 'duration' ? normaliseDuration(actualCurrentValue) : normaliseValue(actualCurrentValue);
+
+    // Detect changes by comparing normalized values
+    let hasChange = !isEqual(normalisedStoredValue, normalisedCurrentValue);
+
+
+    // Add logic for species-based dynamic fields
+    const species = values?.species || [];
+    if (species.length === 0) {
+      
+    } else {
+        species.forEach(speciesName => {
+            const fieldName = `reduction-quantities-${speciesName}`;
+            const speciesStoredValue = values.storedValue?.[fieldName] || null;
+            const speciesCurrentValue = values[fieldName] || null;
+
+            if (!isEqual(normaliseValue(speciesStoredValue), normaliseValue(speciesCurrentValue))) {
+
+              // any change in dynamic fields sets hasChange to true
+                hasChange = true; 
+            } else {
+                
+            }
+        });
+    }
+
+    return hasChange;
```
```suggestion
    if (!isEqual(normalizedStoredValue, normalizedCurrentValue)) {
      return true
    }

    const hasSpeciesChange = (speciesName) => {
      const fieldName = `reduction-quantities-${speciesName}`;
      const speciesStoredValue = values.storedValue?.[fieldName] || null;
      const speciesCurrentValue = values[fieldName] || null;

      return !isEqual(normalizeValue(speciesStoredValue), normalizeValue(speciesCurrentValue));
    }

    const species = values?.species || [];
    return species.find(species => hasSpeciesChange(species)) != null;
```

Also future work: create a ticket to move field specific overrides like this to the schema for that field.

mohomeoffice replied:
Code refactored

jeff-horton-ho-sas commented on client/pages/sections/protocols/sections.js:
```diff
@@ -106,13 +110,68 @@ const getBadges = (section, newComments, values) => {
+  const fieldsWithoutValues = [];
+
+  section.fields?.forEach((field, index) => {
+    console.log("Processing field:", field.name);
```
These logs are going to be very noisy in opensearch when doing server-side rendering of the page, and could potentially be logging PII which would be a security risk
```suggestion
```

jeff-horton-ho-sas reviewed (changes requested):
> It would be good to see some test cases added for hasDatabaseChanges, hasSectionChangedDeep, and hasChangedFields. There are some existing tests in https://github.com/UKHomeOffice/asl-projects/tree/master/test/specs as an example.

I'd still like to see some unit tests for the main change detection functions.

> It looks like there is a bug in eslint causing the linting to fail (https://drone-gh.acp.homeoffice.gov.uk/UKHomeOffice/asl-projects/4773/1/3). It looks like this might be fixed in more recent versions https://github.com/eslint/eslint/issues/14172

The eslint issue still needs fixing

kiran-varma-home-office commented on client/components/editor/index.js:
```diff
@@ -205,4 +205,4 @@ class TextEditor extends Component {
 }
 
-export default connect(null, { throwError })(TextEditor);
+export default connect(null, { throwError })(TextEditor);
```
No change in this file, please revert it to master.

kiran-varma-home-office commented on client/components/checkbox.js:
```diff
@@ -19,9 +19,9 @@ const NtsCheckBoxWithModal = (props) => {
 
-    if (itemRemoved) {
-      const {hasData} = hasExistingDataForCheckbox(project, checkboxValue);
+    const {hasData} = hasExistingDataForCheckbox(project, checkboxValue);
```
@mohomeoffice - I have commented on the PR you closed but this is still not addressed. Could you please check all comments from other PR?

I don't see a reason why hasData scope is changed?

kiran-varma-home-office replied:
If you revert this, there is no change in this file and can be reverted to master.

kiran-varma-home-office commented on client/components/review-field.js:
```diff
@@ -296,4 +296,4 @@ const mapStateToProps = ({ project, settings }, props) => {
 };
 
-export default connect(mapStateToProps)(ReviewField);
+export default connect(mapStateToProps)(ReviewField);
```
No change in this file, please revert it to master.

mohomeoffice replied:
esLint is giving an error due to the absence of a newline at the end of the file.

kiran-varma-home-office commented on client/components/review.js:
```diff
@@ -107,21 +204,32 @@ class Review extends React.Component {
     isGranted,
-    previousProtocols
+    previousProtocols,
+    storedValue, // safely provided stored value
```
Comment doesn't seem to add any value

kiran-varma-home-office commented on client/components/static-section.js:
```diff
@@ -69,4 +69,4 @@ const mapStateToProps = ({ project, application: { isGranted, schemaVersion, isF
 };
 
-export default connect(mapStateToProps)(StaticSection);
+export default connect(mapStateToProps)(StaticSection);
```
No change in this file, please revert it to master.

kiran-varma-home-office commented on client/pages/sections/protocols/protocol-sections.js:
```diff
@@ -186,4 +186,4 @@ class ProtocolSections extends PureComponent {
 const mapStateToProps = ({ application: { schemaVersion }, project }) => ({ schemaVersion, project });
 
-export default withRouter(connect(mapStateToProps, { keepAlive })(ProtocolSections));
+export default withRouter(connect(mapStateToProps, { keepAlive })(ProtocolSections));
```
No change in this file, please revert it to master.

kiran-varma-home-office commented on client/pages/sections/protocols/sections.js:
```diff
@@ -21,7 +21,11 @@ import Conditions from '../../../components/conditions/protocol-conditions';
 import ChangedBadge from '../../../components/changed-badge';
 import {reusableStepFieldKeys} from '../../../helpers/steps';
 
+
+
```
clean up please

kiran-varma-home-office commented on client/pages/sections/protocols/sections.js:
```diff
@@ -106,13 +110,65 @@ const getBadges = (section, newComments, values) => {
+        value: fieldValue,
+      });
+    } else {
+      fieldsWithoutValues.push({
```
I don't see `fieldsWithoutValues` used anywhere, so why do we need it?

kiran-varma-home-office commented on client/pages/sections/protocols/steps.js:
```diff
@@ -556,4 +556,4 @@ export default function Steps({project, values, ...props}) {
     </div>
   );
-}
+}
```
No change in this file, please revert it to master.

kiran-varma-home-office commented on client/reducers/changes.js:
```diff
@@ -39,4 +39,4 @@ const changes = (state = INITIAL_STATE, action) => {
 };
 
-export default changes;
+export default changes;
```
No change in this file, please revert it to master.

kiran-varma-home-office commented on client/components/application-summary.js:
```diff
@@ -20,6 +20,48 @@ import Submit from './submit';
+ * @param {any} value - The input value to normalise.
+ * @returns {string} - A normalised string representation of the input value.
+ */
+function normaliseValue(value) {
```
As suggested previously in other PR, I find it easier to read and write tests when functions like this can be extracted out to different file. This file is too long for my liking :)

kiran-varma-home-office commented:
@mohomeoffice please let me know if you need help with all the cleanup/revert comments. It shouldn't take more than few minutes to fix. I can do it for you and push a commit.

mohomeoffice commented on client/components/application-summary.js:
```diff
@@ -20,6 +20,48 @@ import Submit from './submit';
+
+return String(value).trim(); // Convert other types to string and trim whitespace.
+
+}
```
code refactored

mohomeoffice commented:
Hi Guys, Thank you reviewing my code 

Summary of code pushed : 

1. Fixed ESLint issues.
2. Refactored section, field change detection. ( placed normalisation, duration, field change detection, section detection  code as helper function to keep current code clean ) 
3. Updated test cases. ( created test cases for both fields detection and section change detection ) 
4. Removed unnecessary logs and spaces. (refactored number of nested checks ) 
5. Reverted unchanged files. ( not all of them as after removing space end of line eslint was giving error ) 

Please review and let me know if anything else is required.

Thank you 
Mo

kiran-varma-home-office commented:
Looking good from a quick check. I don't have time to check properly this afternoon. 

Checks are failing, please fix.

farrukhishaq commented on client/helpers/section-change-detection.js:
```diff
@@ -0,0 +1,35 @@
+import isEqual from 'lodash/isEqual';
```
This function could benefit from refactoring making the logic more modular.

mohomeoffice replied:
Thank you and I will try to make it modular at second part of the task

jeff-horton-ho-sas reviewed (approved):
Please update the package version before merging.
